packages/providers/s3/__test__/s3-provider.test.ts

```diff
@@ -0,0 +1,170 @@
+import { S3Provider, S3ProviderConfiguration } from '../src'
+import { FileStatus } from 'shared-types'
+import { S3KeyResolvers } from '../src/utils/s3-key-resolver'
+import { afterEach } from 'node:test'
+import { resetS3Mock } from './s3-mock/localstack'
+
+const expectS3Url = (
+    url: URL,
+    path: string,
+    fileId: string,
+    isUpload = true
+) => {
+    const s3PathResolver = new S3KeyResolvers({
+        uploadBucketPath: path,
+        resourceBucketPath: [path],
+    })
+
+    expect(url.protocol).toBe('http:')
+    expect(url.hostname).toBe('test-bucket.s3.localhost.localstack.cloud')
+    expect(url.port).toBe('4566')
+
+    const pathname = url.pathname.slice(1)
+
+    if (isUpload) {
+        expect(pathname).toBe(s3PathResolver.upload.resolve(fileId))
+    } else {
+        expect(s3PathResolver.resource.resolve(fileId)).toContain(pathname)
+    }
+
+    const searchParams = url.searchParams
+
+    expect(searchParams.get('X-Amz-Algorithm')).toBe('AWS4-HMAC-SHA256')
+    expect(searchParams.get('X-Amz-Content-Sha256')).toBe('UNSIGNED-PAYLOAD')
+
+    const credentialDateString = new Date()
+        .toISOString()
+        .replace(/T.*/g, '')
+        .replace(/[:-]/g, '')
+
+    expect(searchParams.get('X-Amz-Credential')).toBe(
+        `test/${credentialDateString}/eu-central-1/s3/aws4_request`
+    )
+    expect(searchParams.get('X-Amz-SignedHeaders')).toBe('host')
+}
+
+const uploadFile = async (url: string, file: Blob) => {
```

Comment:
This one too. Abstracting it in a separate file would make this file more clear.